### ISSUE TITLE
VuexFire 2

### ISSUE THREAD

**Issue by posva** [posva/vuexfire#14]
Following recent issues #12, #13 and #9, I've been thinking about how vuexfire API should be in the future. Now, I know the v1 is not even out but the idea is to make the current version a v1 with the least possible amount of breaking changes and adding deprecations for v2.
In v2, support for Vue 1, Vuex 1 and Firebase 2 will be dropped while v1 would still be supporting them all.

## What is wrong with the current API

There are many things that aren't currently optimal and that I'd like to address in v2

### Double efforts to bind a reference

ATM you have to declare a variable in the state and also tell vuexfire what's the firebase reference that should go there:

``` js
// store.js
const state = {
  items: null // binded later
}
```

``` js
// some-view.vue
firebase: {
  items: db.ref('items')
}
```

**This is actually not needed** because I can just do `Vue.set(state, key, value)` and the data will be reactive, which was the problem I was facing in the first time. I tried it and works fine at least for Vue 2.
However, **this also makes it harder to see how the state is built**, since you cannot just check `store.js` (or any other module) and walk through the `state` variable. Your state is spread all around your app, which is not convenient.
I therefore think making the user declare the state is a good thing but I should not force that unless it is the only way.

### Using `firebase` option makes no sense

Initially the `firebase` option was taken from vuefire, however, it doesn't make sense having to use a Vue instance to bind the references to the Vuex store.
In the current state, it's very easy to migrate from vuefire to vuexfire but **you cannot use both at the same time** which makes sense in very specific situations.

Binding, unbinding should be doable out of a vm instance. In other words vuexfire should expose functions for this purpose and **not add** any extra method to the Vue prototype.

## What I think would be a good new API for vuexfire

Removing the `firebase` option to declare references binding alongside all the methods/variables added to the Vue prototype (`$bind`, `$unbind`, `$firebaseRefs`).
This will allow `vuefire` to work at the same time that vuexfire.

Installing VuexFire as usual:

```js
// main.js
Vue.use(Vuex)
Vue.use(VuexFire)

// store.js
import { generateMutations } from 'vuexfire'

var sore = new Vuex.Store({
  state: {
    list:  null
  },
  mutations: {
    ...mutations,
    ...generateMutations() // Pass the module name as an argument
  }
})
```

Exposing those as methods:

``` js
import { bind, refs } from 'vuexfire'

setListRef ({ commit, state }, { listRef }) {
  // The current state may have another ref binded
  // state === { list: [...] } 
  // the bind method needs access to the commit and a key to bind on the state
  const unbind = bind(commit, 'list', listRef)
  // VuexFire will internally get the path from the ref and save it
  // so you can later on access it

  // call unbind() if you manually want to unbind
}

// You still don't need an action but here's an example
// using one
addToList (context, { item }) {
  // Giving the path of the ref was 'items'
  // It could have been 'users/someUser/list'
  refs('list').push(item)
}
```

Notice how you no longer specify the module you're binding to because the bind method use a commit function that is bound to the module you're working with. Internally a reference to the commit function is saved so calling `bind` again on the same key within the same module will unbind a previously bound reference.

Since VuexFire is no longer aware of the module you're working in, you cannot use a method like `unbind('cart.items')` to unbind the key `items` of the module `cart`. This should not be a problem since unbinding is done automatically when binding a new value and since you can always dispatch an action to call the `unbind` method.

If the user **manually register** any module then, he would also have to **manually unbind** any refs used by that module when unregistering it.

Please, share your thoughts on this :smile:

**Comment by chrisbraddock** [comment]
Double efforts to bind a reference - I'd rather see this maintained in one place (store). I don't want this -> "this also makes it harder to see how the state is built". I'd vote for clarity over convenience.

I don't understand why someone would use Vuefire and Vuexfire at the same time. The thought of it hurts my head. :)

**Comment by posva** [comment]
Thanks for the feedback!
Defining it on the store is not enough because you always have to bind the state to a reference that is only available at runtime.

> I don't understand why someone would use Vuefire and Vuexfire at the same time. The thought of it hurts my head. :)
I wouldn't promote that anyway 😆

**Comment by zoomclub** [comment]
Glad to hear about the progress, I'll spend sometime checking out the Firebase API and reviewing my use case for Vuexfire, then try to provide some feedback during a busy season here. 

Also, if the readme can clarify more about why Vuexfire decided to do things the way it does that would be helpful. For instance, why not just use the Firebase API directly in Vuex action, mutate and getter methods?

Overall I'm in favor of keeping the binding between Vuex and Firebase as transparent as possible. The mapping between Vue components and Vuex modules hits the sweetspot and hopefully the same kind of elegance can be achieved in Vuexfire.

**Comment by chrisbraddock** [comment]
Here's how I've [attempted to use Vuex with modules](https://github.com/chrisbraddock/vuefire-auth-demo/blob/feature/vuexfire-actions-modules/index.html#L41) today if it tells you anything. :)

**Comment by posva** [comment]
> Also, if the readme can clarify more about why Vuexfire decided to do things the way it does that would be helpful. For instance, why not just use the Firebase API directly in Vuex action, mutate and getter methods?

Because that wouldn't make things reactive 😄 With vuexfire, any client changing a property will sync across all active users

> Overall I'm in favor of keeping the binding between Vuex and Firebase as transparent as possible. The mapping between Vue components and Vuex modules hits the sweetspot and hopefully the same kind of elegance can be achieved in Vuexfire.

Hmm, maybe something like `mapActions` to automatically generate the actions that set a reference:

``` js
var module = {
  state: { cart: null },
  getters: {...},
  actions: {
    ...actions,
    ...mapActions(['cart'])
    // Will add an action named setCartRef(firebaseRef)
    // Or ...mapActions({ setCart: 'cart' }) to have setCart as the 
    // action name instead
  }
}
```

@chrisbraddock Thanks for sharing, I took a look and seems good 👍

**Comment by DevEclipse** [comment]
can the firebase option in my VUE instance ommited and put those pushing update and delete in vuex?

for Example i want to put this on

the actions of the store rather than the vue instance

    addUser(user) {
      this.addProfile(user,'manager');
      this.addProfile(user,'employee');
      this.addProfile(user,'customer');
      this.$firebaseRefs.users.child(user.username.toLowerCase()).set(user);
    },

and in the store

il just create ADD_USER mutation and 

dispatch it from an action

**Comment by posva** [comment]
@DevEclipse It looks like what I explained about setting refs in the original post, so yes 😄

**Comment by DevEclipse** [comment]
keep it up hope you finish vuexfire 2 :)

its really better to make of use actions of vuex

and just use keys for manipulating the database

**Comment by DevEclipse** [comment]
Offtopic:

do i really need to declare this to every module?
Example : Stores Module
mutations: VuexFire.moduleMutations('stores');

what is the difference between this to that?

VuexFire.mutations

cant i just declare it to the main store? or maybe do a mapping?

**Comment by DevEclipse** [comment]
is this a right setup?

```
import VuexFire from 'vuexfire'
import {db,firebase} from '../../firebase_config'
import * as types from '../types'

const state = {
  users: null,
  user: null,
  current_user: null,
};

const getters = {
  getAuth: function (state) {
    return state.users
  }
};

const mutations = Object.assign({},
    VuexFire.moduleMutations('users'),
    {
      [types.ADD_USER](state,user) {

      },
      [types.DELETE_USER](state,user) {

      },
      [types.UPDATE_USER](state,user) {

      },
      [types.GET_USER](state,user) {

      },
      [types.CURRENT_USER](state,user) {

      },
    })
export default {
  state,
  getters,
  mutations
}
```

**Comment by DevEclipse** [comment]
```
import VuexFire from 'vuexfire'
import auth from './modules/auth'
import users from './modules/users'
import items from './modules/items'
import stores from './modules/stores'
import managers from './modules/managers'
import employees from './modules/employees'
import customers from './modules/customers'
import transactions from './modules/transactions'
import getters from './getters'

export default {
  modules: {
    stores,
    items,
    users,
    managers,
    employees,
    customers,
    transactions,
    auth,
  },
  mutations: VuexFire.mutations,
  getters,
}
```

Do i really need VueFire.moduleMutations?

It works just the same

Look Here

![image](https://cloud.githubusercontent.com/assets/19868205/20563430/94d9503e-b1c3-11e6-82c5-1005036fe1bc.png)

**Comment by zoomclub** [comment]
Looking only at Firebase and it's usage in my app I'm going to need a way to create/register and remove/unregister dynamic refs per store module. In my scenario refs are always changing (de/activation) so being able to maintain and track them without simply hardcoding them is essential. Is this how Vuexfire works?

**Comment by chrisbraddock** [comment]
@zoomclub you can do that now with `$bindAsArray` / `$bindAsObject`. There's an [example here](https://github.com/chrisbraddock/vuefire-auth-demo/tree/feature/vuexfire-actions).

**Comment by posva** [comment]
@DevEclipse That's weird, it looks as if you're committing root level mutations only otherwise they are prefixed by the name of the module. But please don't hijack this issue 😛

**Comment by DevEclipse** [comment]
@chrisbraddock in your example when i use it on modules do i need to call the root vue instance for persisting data?
like this?
For example i have a component named message not on the root instance of the vue
then ill do something like this?

MessageMyFirebase Component 

        sendMessage: function ( title ) {
          this.$store.dispatch('sendMessage', {
            messagesRef: this.$root.$firebaseRefs.messages,
            title: title
          })
          this.newMessageTitle = ''
        },

like that?

@posva can't wait for the firebaseRef in the store instance :D

**Comment by chrisbraddock** [comment]
@DevEclipse please open a question on that repo if you're still having issues. This thread is veering off topic.

**Comment by zoomclub** [comment]
@chrisbraddock Thanks for pointing out the `$bindAsArray / $bindAsObject` functions and how they are used in the linked example. This does tie into where I'm getting to as I review my scenario.

@posva I'm on limited time and will aim to clarify my use case over the coming weekend. Presently I'm just reviewing Firebase, my data sources and sinks and how they connect in the context of Vuex. Once that is refreshed to me I look forward to sharing my requirements and learning more about how VuexFire applies.

**Comment by chrisbraddock** [comment]
@posva re: double-binding. What I'm doing today is ... I don't use the `firebase` property on views at all. I'm storing all references on module state and set references only via Vuex `actions` and `mutations`. Actions allow me to handle async in the case where I might have to grab a reference or value before being able to set a 'dependent' reference. Of course this is all chainable with promises.

Since `actions` always receive `rootState`, in `main.beforeCreate()` I commit mutations to store `{ main: this, db: firebase.database() }` on `rootState`. Hence I always have `main` and `db` in `actions` to pass to `mutations` so I can `main.$bindAsArray|Object .... db.ref()`.

I need to prove it out a little more but so far so good, fingers crossed.

**Comment by zoomclub** [comment]
I'm reporting in to say I've learnt a few new things concerning my use case for firebase. I needed a firebase refresher and now it is easier to see what role vuexfire plays. Essentially, it binds firebase refs to store module state through a set of generic ops that correspond to the typical handful of sink ops firebase provides. IMO, vuexfire is presently not as clear of an API as it could be but this could change.

New insights into firebase have allowed me to reduce the number of store modules I require and let me map multiple JSON collections onto them. I now have a good idea about what the firebase endpoints in each store module will be, so it likely just makes sense to declare the endpoint refs as sources and sinks right in the store modules. I would do this right in state except I want offline first with local storage so I'd declare them outside of state and then set state props from there.

Instead of just receiving sources when changes are made and setting store module state I'm kind of thrilled by the option of RxJS observables. So for sources the inbound refs can be turned into observables that can be processed by RxJS operators before finally setting the result in a store module state. For sinks it should be possible to plug a RxJS stream into a outbound ref and both set what terminal node the ref should target and also send the payload to it.

Its still coming together here but I'm loving how RxJS and firebase refs get along, see more here: https://www.youtube.com/watch?v=tQBHglmiygw&index=12&list=PLdssc-pDiZ7OD78kJVp4habTynj-Etwhm

**Comment by posva** [comment]
@zoomclub Thanks for sharing your thoughts. From what I read it looks like things are going in the right direction for you.
About RxJS, I don't think I'll add direct support for it because that may bloat the codebase but I'll expose the listeners created by firebase so you can plug in and create your own observables.

**Comment by zoomclub** [comment]
@posva Like the idea of being able to plugin RxJS observables 👍  It might be good to have a factory function that creates read and write refs dynamically? Then just by calling the factory refs could be un/registered in any store module.

It would also be useful to be able to change the end terminal node a given read or write ref is pointing to. This might be better as aspect of the factory because refs are easy to define and dispose anyways. So a factory function/ref constructor might just have optional args to pass in like: ref-name, nested-path and terminal-node.

Just thoughts for now, what are your thoughts about such a factory approach?

**Comment by zoomclub** [comment]
Found this offline wrapper for Firebase, could something like this be added to VuexFire?

https://github.com/yeldarby/offline-firebase

**Comment by posva** [comment]
>  It might be good to have a factory function that creates read and write refs dynamically?

Yes, that was the `mapActions`.

I don't know what you're referencing as _terminal-node_
My thoughts on this is that I don't want to introduce a magic api, only helpers that could be written by the user but since they're used so often are worth adding to the lib itself.

About writing to the firebase, I think that should be handled by the user because some may want to use firebase differently

**Comment by posva** [comment]
About the offline wrapper, it looks kind of outdated. Storing in localstorage is something that depend on the application is not always needed, so I think that belongs to the userland

**Comment by zoomclub** [comment]
@posva The terminal-node is just the position in the Firebase tree a given ref is pointing to, it works like a cursor. I'd just like to keep the ref around yet allow what its pointing at to be changeable. Maybe I'm over engineering but that kind of flexibility makes sense to my app.

Just throwing out ideas for the most part. Getting a good idea of what I need now and most of that will be centered around Firebase and RxJS and basic Vuex as far as I can tell.

**Comment by Granipouss** [comment]
Hi ! I find that project quite interesting.
What would you think of an API kind of similar to [vuex-router-sync](https://github.com/vuejs/vuex-router-sync) ? Something like that:
``` js
import { sync } from 'vuexfire'

sync(store, {
  todos: to dos réf
})

// Which create the dB module
store.state.db.todos
// ...
```

**Comment by posva** [comment]
@Granipouss The problem with that syntax is that the you're binding to the store and you want to bind to modules too. I don't want to create a module with all the data, I prefer dividing the data by concern

**Comment by Granipouss** [comment]
I see your point, then maybe give that as an easy way and allow users to do more specific stuffs with more functions ?
A thing I love about  vue is that you can do things easily if you want or go deeper and pretty much whatever you want. I feel like this is something that tend to get lost with plugins. That's why I'didn't propose a very basic api at least as a complement of the more full and complexes one.
(I don't know if I'may very clear with my explanationson, sorry if I wasn't 
^^')

**Comment by posva** [comment]
> to do more specific stuffs with more functions 

No, that's not clear 😆

**Comment by Granipouss** [comment]
Well here is an API with different level of complexity
``` js
// Easyiest way (maybe even too basic)
import { sync } from 'vuexfire'
const storeA = new Vuex.Store({ ... })

sync(storeA, { todos: todosRef })

// Using modules
import { generateModules } from 'vuexfire'
const storeB = new Vuex.Store({
  ... ,
  modules: {
    ...generateModules({ todos: todosRef })
  }
})

// Using modules
import { generateModules } from 'vuexfire'
const storeB = new Vuex.Store({
  ... ,
  modules: {
    ...generateModules({ todos: todosRef })
  }
})

// Decomposing
import { generateMutations, generateActions, generateGetters } from 'vuexfire'
const store = new Vuex.Store({
  state: { ... },
  mutations: {
    ... ,
    ...generateMutations({ todos: todosRef })
  },
  actions: {
    ... ,
    ...generateActions({ todos: todosRef })
  },
  getters: {
    ... ,
    ...generateGetters({ todos: todosRef })
  }
})
```

There you have different levels of complexity for the user. Someone that want to do specific stuff have access to functions that allow them more customability, but those who just want a basic sync between their vuex store and some firebase database can have it in just one simple line.

**Comment by posva** [comment]
Ok, now I understand. That won't do it because sync uses the store and that's too much. I'd rather use a bind inside an action, that's where it belongs. About the helper functions, the getters are unecessary, but the others are intended to be implemented

**Comment by DevEclipse** [comment]
This is how i do it though

```
import VuexFire from 'vuexfire'
import users from './modules/users'
import items from './modules/items'
import stores from './modules/stores'
import managers from './modules/managers'
import employees from './modules/employees'
import customers from './modules/customers'
import transactions from './modules/transactions'
import auth from './modules/auth'
import stocks from './modules/stocks'
import getters from './getters'

const state = {
    refs: null,
    busers: null,
    bitems: null,
    bstores: null,
    bmanagers: null,
    bemployees: null,
    bcustomers: null,
    btransactions: null,
    btags: null,
    bstocks: null,
    bnotifier: null,
    server_time: null,
    updateObject: null,
    newObject: null,
    deleteObject: null,
    list: null,
};
const mutations = {
  ['SET_REFS'](state, payload) {
    state.refs = payload;
  },
  ['SET_UPDATE_OBJECT'](state, object) {
    state.updateObject = object;
  },
  ['SET_NEW_OBJECT'](state, object) {
    state.newObject = object;
  },
  ['SET_DELETE_OBJECT'](state, object) {
    state.deleteObject = object;
  },
  ...VuexFire.mutations
};
const actions = {
  newObject({commit, getters}, object) {
    let resultObject = _.clone(object);
    resultObject['created_date'] = resultObject['updated_date'] = getters.serverTime;
    commit('SET_NEW_OBJECT', resultObject);
  },
  updateObject({commit, getters}, object) {
    let resultObject = _.clone(object);
    resultObject['updated_date'] = getters.serverTime;
    if (resultObject['.key']) {
      delete resultObject['.key'];
    }
    commit('SET_UPDATE_OBJECT', resultObject);
  },
  addRefObject({dispatch, getters}, {ref, value}){
    let data = value || getters[`stored${ref}`];
    dispatch('newObject', data);
    getters[`ref${ref}s`].push(getters.getNewObject);
  },
  updateRefObject({dispatch, getters}, {ref, value, action}){
    let data = value || getters[`stored${ref}`];
    dispatch('updateObject', data);
    getters[`ref${ref}s`].child(data['.key']).update(getters.getUpdatedObject);
    dispatch(action, getters.getUpdatedObject);
  },
  deleteRefObject({commit, getters}, {ref, value}){
    let data = value || getters[`stored${ref}`];
    getters[`ref${ref}s`].child(data['.key']).remove();
    commit('SET_DELETE_OBJECT', data);
  },
};
const modules = {
    users,
    managers,
    employees,
    customers,
    stores,
    stocks,
    items,
    transactions,
    auth,
};

export default {
  state,
  mutations,
  getters,
  actions,
  modules,
}

```

and in my item module


```
const state = {
  storedItem: null,
  foundItem: null,
};

const getters = {
  currentItem(state,getters) {
    if(!getters.allItems) return;
    return _.find(getters.allItems,['.key',getters.routeParams.item]);
  },
  currentItemStocks(state,getters) {
    if(!getters.currentItem) return;
      return _.filter(getters.allStocks,['item',getters.currentItem['.key']]);
  },
  storedItem(state) {
    if(!state.storedItem) return;
    return state.storedItem;
  },
  storedItemStocks(state,getters) {
    if(!getters.storedItem) return;
    return _.filter(getters.allStocks,['item',getters.storedItem['.key']]);
  },
  foundItem(state) {
    if(!state.foundItem) return;
    return state.foundItem;
  }
};

const mutations = {
  ['STORE_ITEM'](state,item) {
    state.storedItem = item;
  },
  ['STORE_FOUND_ITEM'](state,item) {
    state.foundItem = item;
  }
};

const actions = {
  addItem({dispatch},value) {
    dispatch('addRefObject',{ref: 'Item', value});
  },
  deleteItem({dispatch},value) {
    dispatch('deleteRefObject',{ref: 'Item', value});
  },
  updateItem({dispatch},value) {
    dispatch('updateRefObject',{ref: 'Item', value,action: 'storeItem'});
  },
  storeItem({commit,getters},item) {
    commit('STORE_ITEM',_.clone(item) || {
        name: '',
        cost_price: 0,
        image_url: '',
        description: '',
        created_by: getters.authManager['.key'],
        rating: 0,
        tags: ['item'],
      });
  },
  findItem({commit,getters},key) {
    commit('STORE_FOUND_ITEM',_.find(getters.allItems,['.key',key]));
  }
};

export default {
  state,
  getters,
  mutations,
  actions
}
```

getters.js

```
  allItems(state) {
    return state.bitems;
  },
  refItems(state) {
    return state.refs.bitems;
  },

```

hope it helps but its dirty for me

still waiting for VuexFire 2 great work keep it up

**Comment by DevEclipse** [comment]
Created this idea im just passing the $firebaseRefs in beforeCreate in the root vue

and use this to generate my getters and actions

```
const generateGetters = function (refs) {
  let refKeys = Object.keys(refs);
  let getters = [];
  _.forEach(refKeys,key => {
    let capKey = key.charAt(0).toUpperCase() + key.slice(1);
    getters[`all${capKey}`] = function(state) {
      return state[`b${key}`];
    };
    getters[`ref${capKey}`] = function(state) {
      return state.refs[key];
    }
  });
  return getters;
};
const generateActions = function (refs) {
  let refKeys = Object.keys(refs);
  let actions = [];
  _.forEach(refKeys,key => {
    let capKey = key.charAt(0).toUpperCase() + key.slice(1);
    let singularCapKey = capKey.slice(0, -1);
    actions[`add${singularCapKey}`] = function({dispatch},payload) {
      dispatch('addRefObject',{ref: singularCapKey, payload});
    };
    actions[`delete${singularCapKey}`] = function({dispatch},payload) {
      dispatch('deleteRefObject',{ref: singularCapKey, payload});
    };
    actions[`update${singularCapKey}`] = function({dispatch},payload) {
      dispatch('updateRefObject',{ref: singularCapKey, payload});
    }
  });
  return actions;
};
```

**Comment by posva** [comment]
@DevEclipse I deleted your comments with code because they were off topic and taking up too much place, making it hard to follow the conversation. I hope you don't mind

**Comment by zoomclub** [comment]
Hi @posva, happy new year! Just musing about Vuex, Firebase and RxJS again, this time considering a new wrapper API for Firebase, which can be found here: https://github.com/jsayol/rxfirebase

The rxfirebase wrapper looks good and will cover the spectrum of Firebase services by the looks of it, plus it is also TypeScript ready. To be fair I find the current Vuexfire API difficult to apply to my use case and so having a separate wrapper clarifies things for me at this point. 

Just thoughts but why not just have a separate object that uses RxFirebase for all Firebase editing and querying? So all the active refs plus custom edit and query functions to Firebase would live in this hub object. This object would then also be directly invoked from components, without having to engage Vuex.

Next, the RxJS observables in this hub object can then be processed to produce state to pass to Vuex actions in store modules. Another exciting option would be to just pass whole observables to vuex-observable, which lets you "Consume Vuex actions as Observables using RxJS 5". See https://github.com/vuejs/vuex-observable

So, vuex-observable is the connection between the hub object and Vuex to produce state via store module actions fanning out to mutations, sound good? All the standard Vuex to Vue component mapping still applies and Firebase, RxJS and Vuex all get along in a clear cut way.

**Comment by posva** [comment]
Hi @zoomclub, happy new year to you too.
Because you don't need RxJS. VuexFire must be agnostic and light and **use vuex**. RxJS can be amazing for some use cases but not everyone needs it.

**Comment by caex** [comment]
Hello, thanks for spending time to this project,

I just tested the v2 branch, read the example at https://github.com/posva/vuexfire/blob/v2/examples/TodoApp/modules.html and WOW 10 minutes later it's integrated in my app 😮 🎉 

I can use the `Firebase.auth().onAuthStateChanged({uid})` event to dispatch a `bindCart (uid)` action that will simply ````bind('cart', db.ref(`carts/${uid}`))````

Now my [cart] store module is always in sync depending on the uid of the user logged in. What sounded like a probable headache is in the end crystal clear.
Thats awesome ! thanks for this plugin !!

**Comment by posva** [comment]
@caex Cool! Glad you can test it before the release! If you have any other thoughts, make sure to share 😉

**Comment by chrisbraddock** [comment]
@posva I'm Planning to test this out today as well and I'm hoping to write some deployable code around it. You've mentioned some broader changes possibly involving Vuex being discussed. Can you supply any perspective on where the project is headed in the near future? Are there any particular relevant Vuex issues worth watching?

**Comment by posva** [comment]
The relevant part that may change is how to bind and unbind: instead of generating them (which is actually not handy at all when splitting up files (https://github.com/posva/vuexfire/blob/v2/test/bind-array.spec.js#L36)). They should be available directly into the context so you can write something like this: 
```js
someAction ({ commit, bindFirebaseRef }, { ref }) {
  bindFirebaseRef('key', ref)
}
```
But we're still discussing the feature and I have yet to find the time to fully implement it (https://github.com/vuejs/vuex/pull/571)

Other than that, it shouldn't change much. The core will probably stay the same.
If you find any issues or have thoughts, make sure to share them! 🙂

**Comment by chrisbraddock** [comment]
@posva thanks - will do. I only recently became aware of [Firebase multi-path updates](https://www.youtube.com/watch?v=i1n9Kw3AORw&t=356s). Because of the denormalized nature we need to implement ... I guess it sounds appealing and I'm assuming a best practice at this point. I'm wondering if there are considerations with regards to VuexFire.

**Comment by posva** [comment]
@chrisbraddock No, because VuexFire doesn't control how you write to the database, it only reads from it

**Comment by chrisbraddock** [comment]
@posva Damn, I wish I had made that distinction before. It's more obvious now that you pointed it out, and should have been without it. :) It definitely helps me having a clearer idea of VuexFire concerns as I fumble around.

I'm trying to implement the binding you setup w/ `generateBind()` but it hasn't taken me long to get a little confused.

I want to use a [reusable table component](http://element.eleme.io/#/en-US/component/table) that has a prop named `data` with `type` `Array`. However, my data in this case is stored as an object and [I don't think I should change it](https://firebase.googleblog.com/2014/04/best-practices-arrays-in-firebase.html).

It looks like other libraries have [handled this in the past](https://firebase.googleblog.com/2014/05/handling-synchronized-arrays-with-real.html). I'm not sure if using your `bindAsArray` thrown at an object, might provide this kind of functionality. However I think `generateBind` uses its own introspection to decide how to bind, so that option isn't available. Thoughts?

**Comment by posva** [comment]
> I'm trying to implement the binding you setup w/ generateBind() but it hasn't taken me long to get a little confused

I'm not surprised, that function is a bad piece of API 😕. Basically, it takes either the store or a module: `const { bind, unbind } = generateBind(t.context.store._modules.get(['todos']))`. They can then be used anywhere. You probably need a layer of abstraction for that until the new plugin API for vuex is finalised and merged.

> However I think generateBind uses introspection to decide how to bind, so that option isn't available. Thoughts?

It's actually `bind` that does that. It checks if the value in the state is an array to bind it as an array. Otherwise, it binds it as an object. I found that more intuitive that having 2 different functions. The check is done [here](https://github.com/posva/vuexfire/blob/v2/src/index.js#L136-L140)

> However, my data in this case is stored as an object and I don't think I should change it based on my use case and Firebase best practices for arrays.

I'm not sure to follow because everything is an object in Firebase. If your object is actually a collection of subobjects, it can probably be retrieved as an array

**Comment by chrisbraddock** [comment]
> I'm not sure to follow because everything is an object in Firebase. If your object is actually a collection of subobjects, it can probably be retrieved as an array

Sorry, I wasn't being specific enough. I could alter my storage schema such that Firebase would **render the object data as an array** that I could pass without modification to the component prop. That's my understanding, anyway based on this from the best practices link from earlier:

> Firebase has no native support for arrays. If you store an array, it really gets stored as an "object" with integers as the key names.
>
> However, to help people that are storing arrays in Firebase, when you call .val() or use the REST api to read data, if the data looks like an array, Firebase will render it as an array.

Per my understanding of that link, however, we really don't want to do this as it busts conflict resolution if multiple clients are attempting to write child data to the "array".

I might have been making some bad assumptions about the relationship between arrays in Firebase, as described above, and `$bindAsArray`.

**Comment by posva** [comment]
I don't know how the data looks like so I may be wrong, but it should work nicely if you bind it as an array 🙂 . As long as you declare it as an array in the state before binding it, you should be good to go

**Comment by chrisbraddock** [comment]
> As long as you declare it as an array in the state before binding it

That was the piece that needed to click ;) I was leaning towards this based on your last message, just hadn't tested yet. Thanks!

BTW, Having some binding mechanism injected in to the context is going to be super nice.

I do think the project is worth your effort for sure - thanks for doing it. Firebase seems to round out the Vue stack pretty darn nicely.

**Comment by posva** [comment]
> Firebase seems to round out the Vue stack pretty darn nicely

I'm do agree so much with that!

**Comment by robbiemu** [comment]
`sync(storeA, { todos: todosRef })` could be doable if you changed it slightly, like:

```js
import VuexFire from 'vuexfire'

VuexFire.sync(storeA, VuexFire.presets.MODULE_SYNC, { todos: todosRef })
```

but anyway, the rest looks beautiful in that example

**Comment by ralphchristianeclipse** [comment]
@robbiemu you can use this https://github.com/ralphchristianeclipse/vuex-firebase

**Comment by posva** [comment]
@robbiemu I forgot to answer this. We still haven't advanced on Vuex counterpart so it haven't moved much...  But the API is quite different from the one in the first post:

```js
new Vuex.Store({
  actions: {
    setTodos ({ bindRef }, { ref }) {
      return bindRef('todos', ref)
    }
  },
  getters: {
    todos: state => state.todos.filter(...)
  },
  plugins: [Vuexfire]
})
```

The idea is to use Vuexfire as a Vuex plugin and it gives you access to `bindRef`, `unbindRef` inside of actions. It makes it usable in modules as well

PS: Removed the spam comment 😆

**Comment by 37ch4** [comment]
This looks great! Can't wait for V2.

**Comment by robbiemu** [comment]
No problems.. You didn't remove any comments of mine, I only had 1 here. Must've been someone else. 

That syntax only addresses getters, but it looks good.. Can't wait

**Comment by posva** [comment]
It wasn't yours 😄 
I'll soon put the v2 branch as the default and release some alphas but I still need to figure out about extending Vuex plugin interface with Evan. Other things were more important so we forgot about this a bit 😣

**Comment by posva** [comment]
VuexFire 2 is released 🎉
If you find bugs, make sure to open a new issue 👍
The API doc is really short, more docs may come with time, see #28 
Thanks to all who gave feedback about the API. In the future, the action enhancer may be able to be applied globally but this is something we'll look on Vuex end ofc.

@chrisbraddock and others who were using the old v2. To migrate from what existed previously: bind and unbind have been renamed to a bit more verbose versions `bindFirebaseRef` and `unbindFirebaseRef`, and are directly accessible in action enhancers with `firebaseAction`. It's not longer a plugin and you only need to provide the `firebaseMutations` to the rootStore (it's actually just one mutation)

**Comment by robbiemu** [comment]
Hey congradulations! I'll do my next fb project with this!

**Comment by chrisbraddock** [comment]
@posva I can't wait go get back to the project using this and try it out! Thanks for your continued work and getting a solution for this pushed through w/ the team!

**Comment by posva** [comment]
@chrisbraddock Thanks! I'm also eager to get some feedback about VuexFire 2 and make improvements
